### PR TITLE
    cache scope and place on CinnLaunchContext and pass them to callback

### PR DESCRIPTION
**Pull request by CtfGo**
### PR types

Function optimization
### PR changes

OPs
### Describe

cinn_launch_op：   cache scope and place on CinnLaunchContext to skip duplicate alloc/free callback construction

cinn_launch_op_test passed:
image

### CODE REVIEW

**Review by zhhsplendid — APPROVED** [review]
LGTM

File: paddle/fluid/framework/paddle2cinn/cinn_compiler.h [hunk 1]
```diff
@@ -31,6 +31,13 @@
 #include "paddle/fluid/platform/macros.h"
 
 namespace paddle {
+
+namespace operators {
+namespace details {
+class CinnLaunchContext;
+}
+}
```
Comment (wzzju) [inline paddle/fluid/framework/paddle2cinn/cinn_compiler.h:39]: }  // details
}  // operators

File: paddle/fluid/operators/cinn/cinn_launch_op.h [hunk 2]
```diff
@@ -116,56 +103,54 @@ class CinnLaunchOpKernel : public framework::OpKernel<T> {
         compilation_key, inputs_name2tensor, target, stream);
     details::DebugCinnCompiledResult(cinn_compiled_object);
 
-    auto launch_context = std::make_unique<details::CinnLaunchContext>(
-        cinn_compiled_object.paddle2cinn_varmap, cinn_compiled_object.scope);
-
+    const auto& launch_context = cinn_compiled_object.launch_context;
```
Comment (wzzju) [inline paddle/fluid/operators/cinn/cinn_launch_op.h:106]:
```suggestion
    auto* launch_context = cinn_compiled_object.launch_context.get();
```

File: paddle/fluid/operators/cinn/cinn_launch_context.cc [hunk 3]
```diff
@@ -32,7 +32,30 @@ CinnLaunchContext::CinnLaunchContext(
...
+                                          const platform::Place& place) {
+  if (std::addressof(scope) == cached_scope_ &&
+      std::addressof(place) == cached_place_) {
+    VLOG(4) << "Captured scope:" << cached_scope_ << ", place:" << cached_place_
+            << " are not changed";
+    return;
+  }
+  cached_scope_ = std::addressof(scope);
+  cached_place_ = std::addressof(place);
+  cached_temp_scope_ = scope.NewTmpScope();
```
Comment (wzzju) [inline paddle/fluid/operators/cinn/cinn_launch_context.cc:45]: 感觉这句后面还应该加上如下两行代码：
```cpp
hold_buffers_.clear();
name2argument_.clear();
```

File: paddle/fluid/operators/cinn/cinn_launch_context.h [hunk 4]
```diff
@@ -40,16 +40,22 @@ class CinnLaunchContext {
...
+  // by callback of execution arguments
+  void UpdateCapturedEnv(const framework::Scope& scope,
+                         const platform::Place& place);
+
+  // Return whether execution arguments has been initialized
+  bool IsArgumentsInitialized() const;
+
   // Return whether a Paddle variable used on compiled kernels
-  bool IsVariableUsed(const std::string& var_name);
+  bool IsVariableUsed(const std::string& paddle_name) const;
```
Comment (wzzju) [inline paddle/fluid/operators/cinn/cinn_launch_context.h:52]:
```suggestion
  bool IsVariableUsed(const std::string& paddle_var_name) const;
```

File: paddle/fluid/operators/cinn/cinn_launch_context.h [hunk 5]
```diff
@@ -40,16 +40,22 @@ class CinnLaunchContext {
...
 
   // Assign tensor buffer to input or output variables
-  void AssignExternalVariable(const std::string& var_name,
-                              const platform::Place& place, LoDTensor* tensor);
+  void AssignExternalVariable(const std::string& paddle_name);
```
Comment (wzzju) [inline paddle/fluid/operators/cinn/cinn_launch_context.h:55]:
```suggestion
  void AssignExternalVariable(const std::string& paddle_var_name);
```

File: paddle/fluid/operators/cinn/cinn_launch_context.h [hunk 6]
```diff
@@ -40,16 +40,22 @@ class CinnLaunchContext {
...
 
   // Assign tensor buffer to internal variables
-  void AssignInternalVariable(const std::string& var_name,
-                              const platform::Place& place, LoDTensor* tensor);
+  void AssignInternalVariable(const std::string& cinn_name);
```
Comment (wzzju) [inline paddle/fluid/operators/cinn/cinn_launch_context.h:58]:
```suggestion
  void AssignInternalVariable(const std::string& cinn_var_name);
```

File: paddle/fluid/operators/cinn/cinn_launch_context.h [hunk 7]
```diff
@@ -72,16 +74,15 @@ class CinnLaunchContext {
...
-  // Share the buffer of a Paddle tensor to CINN by delivering memory address
-  // to a cinn_buffer_t object
-  std::unique_ptr<cinn_buffer_t> ShareTensorWithCinnBuffer(
-      const platform::Place& place, bool free_mem_callback, LoDTensor* tensor);
-
-  // Set an argument with (cinn name)->(paddle tensor) pair
-  void SetArgument(const std::string& cinn_name, const platform::Place& place,
-                   bool free_mem_callback, LoDTensor* paddle_tensor);
+  // Set an argument with (cinn name)->(cinn_buffer_t) pair
+  void SetArgument(const std::string& cinn_name,
```
Comment (wzzju) [inline paddle/fluid/operators/cinn/cinn_launch_context.h:78]:
```suggestion
  void SetArgument(const std::string& cinn_var_name,
```

**Review by wzzju — APPROVED** [review]
LGTM. Some details can be improved in the next PR.

File: paddle/fluid/operators/cinn/cinn_launch_context.cc (hunk 3 above)
Comment (CtfGo) [reply paddle/fluid/operators/cinn/cinn_launch_context.cc:45]: callback 里都是用更新后的 scope 来获取tensor，不是像之前传 tensor 指针，所以不用重新构造参数。

File: paddle/fluid/framework/paddle2cinn/cinn_compiler.h (hunk 1 above)
Comment (CtfGo) [reply paddle/fluid/framework/paddle2cinn/cinn_compiler.h:39]: 将在下个 pr 修改

File: paddle/fluid/operators/cinn/cinn_launch_context.h (hunk 4 above)
Comment (CtfGo) [reply paddle/fluid/operators/cinn/cinn_launch_context.h:52]: 将在下个 pr 修改

File: paddle/fluid/operators/cinn/cinn_launch_context.h (hunk 5 above)
Comment (CtfGo) [reply paddle/fluid/operators/cinn/cinn_launch_context.h:55]: 将在下个 pr 修改

**Pull request merged**